Add `edit` button to code viewer

davidvgalbraith opened the pull request:
So you can edit Juttles.

Fixes https://github.com/juttle/juttle-viewer/issues/29

@mnibecker @go-oleg

davidvgalbraith commented:
Having a terrible time making it resizable...

dmehra commented:
For PRs like this it would be great to deploy the app from the branch code to someplace like demo.juttle.io (on the side, of course, not replacing the main one but in another container) so reviewers can easily try out behaviors. Would that be an easy thing to do, given we have a way to build juttle-engine docker container?

mnibecker commented:
Daria you pull Dave's branch and follow along with our wonderful new readme for see Dave's changes locally: https://github.com/juttle/juttle-viewer/blob/master/README.md

mnibecker commented:
A couple of style recommendations:
- The url query parameter should change when we go into edit mode. Ideally the url changes to `/local`
- Can we make the font for the edit box bigger
- I don't think it needs to be resizeable, but it should be dismissable. It should auto size the area of the text
- The editor should expand to the width of the window
- The background color should match the background color of the app. Make the line highlight a light gray so its a little more pronounced (I can help you find the color)
- Place the light gray border underneath the editor
- Replace the view section the editor. (In non local mode it should be read-only).
- The View button should hide the editor
- Change 'new' to 'local'

davidvgalbraith commented:
Updated with Tuesday's suggestions. I think we're getting close!

dmehra commented:
Trying the current version. 

1) Why is the editor box divided into two parts? I think I can only type in the left one, what's the right one for? (Now discovered that typing a long line takes it right past the divider into the right cell, this can't be the intended behavior.) 
Also the right vertical border is not visible without scrolling a little bit.

![image](https://cloud.githubusercontent.com/assets/5817365/13643919/0b06d6ae-e5d8-11e5-847a-bb4f04f7dec4.png)

2) When in full-screen mode, the buttons overhang into the output area:

![image](https://cloud.githubusercontent.com/assets/5817365/13643956/3f46dfea-e5d8-11e5-88f0-06b038addcb4.png)

3) If you edit the code and click Expand without clicking Run first, you end up at a black screen that has nothing except the play/unexpand buttons, that's not ideal UX. Should we make Expand auto-run if there wasn't already a rendered output?

4) Does this not support inputs yet? I tried a program with input, it doesn't get rendered anywhere, if it has a default then the default value is used by the program, and if it doesn't then I get error when the juttle tries to use the null value.

5) If you run a program that hits an error (TYPE-ERROR in my case), the stop button continues to rotate as if the program is still executing. And I can't get it to stop, have to refresh the browser. By the way, after refresh I ended up with this extraneous "Edit" button on the screen. When I clicked on it, it disappeared:

![image](https://cloud.githubusercontent.com/assets/5817365/13644332/10f2e740-e5da-11e5-8d20-6c67e4fb2eda.png)

To see both (4) and (5), try this juttle:
```
input t: text;

emit -from :-1h: -to :now: -every :1m:
| reduce -every :5m: count()
| view timechart -title "My Chart " + t
```

6) The timechart doesn't have the zoom feature, not sure if this is specific to the new viewer, will talk to @go-oleg about that.

mnibecker commented on src/apps/assets/sass/main.scss:
```diff
@@ -82,6 +82,10 @@ $run-menu-height: 65px;
+    .new-btn {
+        margin-left: 20px;
+    }
+
```
do we need this now?

mnibecker commented:
Nice stuff man! Getting there...

mnibecker commented on src/apps/components/juttle-editor.js:
```diff
@@ -0,0 +1,33 @@
+        return (
+            <AceEditor
+              theme="monokai"
+              height="200"
```
can we make the editor expand to the length of the buffer? You can see how we did it in jut-io/product: https://github.com/jut-io/product/blob/master/src/applib/components/editor/editor.jsx

davidvgalbraith replied:
I'm not sure what you mean by this one.

mnibecker replied:
make the height of editor, the height of the text

mnibecker commented on src/apps/assets/sass/components/_font-button.scss:
```diff
@@ -23,6 +23,14 @@
+    width: 100px;
+    font-size: $font-size-large;
+    margin-bottom: 7px;
+}
```
let's make the button float on the right side of the editor, do this instead
- on parent div add a `position: relative`
```sass
.btn-edit {
    position: absolute;
    top: 25px;
    right: 25px;
}
```

dmehra commented:
A note on inputs to (4) above: looks like they only don't get rendered if you type in your own juttle (ie, when `run mode: Not Set` - should that be something like `run mode: Edit` or `run mode: New`?). If you open a file by browsing paths, then inputs get rendered fine.

mnibecker commented on src/apps/assets/sass/main.scss:
```diff
@@ -119,10 +123,9 @@ $run-menu-height: 65px;
-    @include flex-shrink(0);
+    border-style: solid;
+    border-width: 1px;
+    border-color: $gray-light;
```
to be consistent with the other inserts, just do a `border-bottom: 1px solid $gray-dark`

dmehra commented:
There's a problem with tile expanding beyond the viewable area, try the demo program (with `./bin/juttle-viewer -d -j demo.juttle.io`): http://localhost:2001/?path=%2Fnpm_download_counts.juttle
See how the tile is used to output a header of sorts, which is huge and extends out to the right, and you cannot scroll to see all of it.
Ideally an issue like this would be solved by resizing the tile (and making its font smaller) to fit into the viewable area.

mnibecker commented:
@dmehra on your last comment can you post a screenshot of what you're talking about?

mnibecker commented:
@dmehra a lot of your comments are not within the scope of this pr, lets try to focus on what Dave is doing here and create issues for non-edittable juttle items.

dmehra commented:
Posted a screenshot for the unscrollable tile, realized while doing it that my zoom in the browser was set to 120%; at 100% the tile title fits. But if you zoom, then there's no horizontal scroll. Perhaps we can ignore that since it looks/works fine at 100%.

dmehra commented:
@mnibecker, not sure what you mean by "create issues for non-edittable juttle items". I don't know which parts of what I did are out of scope for this PR. feel free to make separate issues where applicable.

mnibecker commented:
The scope of this pr as dave succinctly states at the top is "So you can edit Juttles."

dmehra commented:
All the issues I ran into happened while editing Juttles. I'll leave it up to @davidvgalbraith if they should be fixed on this branch, or filed as separate bugs, should the branch be merged with them.

dmehra commented:
When you open a juttle and click the "Edit" button, the resulting mode is
```
local
run mode: Path
```
Is that right? Seems like it's no longer "Path" if i'm in local editing. For example, I can't refresh the browser and have it reload from the path I had originally opened. (I'd like to generally understand what "run mode" means, please point me at a doc.)

dmehra commented:
I see two issues here as blockers to merge: 
- support for inputs in free-editing mode (note (4) above), and 
- getting out of bad state after you run a juttle with error (currently you have to refresh the browser, as there's no way out; note (5) above).

I will file the button overhang in fullscreen mode separately, so (2) above can be ignored.

Note (1) above was for styling, please consider it here.

Issue (3) above only got created once the editing ability was added, but it's not a blocker.

The "local, run mode: Path" thing would be best resolved here, but also not a blocker.

mnibecker commented on src/apps/components/juttle-editor.js:
```diff
@@ -0,0 +1,33 @@
+import React from 'react';
+
+import AceEditor from 'juttle-react-ace';
```
I'm kind of questioning what we really get for using `react-ace`. Seems like it kind of just obscures a lot the ability to use the full range of customization available on ace. I'm kind of leaning towards just using `brace` instead.

davidvgalbraith replied:
I like `react-ace`. I'd lean towards making further modifications to `juttle-react-ace` if we need any other functionality.

mnibecker commented on src/apps/components/juttle-editor.js:
```diff
@@ -0,0 +1,33 @@
+        };
+        var commands = [RUN_JUTTLE];
+        return (
+            <AceEditor
```
Can we add a scroll margin to the top, so the top line of text isn't so close the border. Using react-ace you would have to add a `ref="aceEditor"` tag to the `<AceEditor />` component, then in `componentDidMount` do something like this:
```javascript
this.refs.aceEditor.renderer.setScrollMargin(15, 0)
```

mnibecker commented:
lets add a file `_editor.scss` to sass/components where we can add some overrides to our monokai styling. The biggest one is making the background-color consistent with with the rest of the app. Should do something like this:

```sass
.ace-monokai {
    background-color: $gray-darker;
}

davidvgalbraith commented:
I fixed the inputs, got rid of the 80-character divider, made the background color match, moved the edit button, and set the scroll margin. Still need to clean up the code a bit and fix the coverage.

dmehra commented:
I'm having trouble using the editor, as i type on say line 4 (adding a line, where a 10-line juttle already exists), it keeps reporting syntax errors while i type, and periodically jumps my cursor to the last character of the whole juttle. It's hard to explain any better, but surely you'll get the same behavior? This is in freewheeling "local" edit mode. See me typing "Don't jump" in there?

![image](https://cloud.githubusercontent.com/assets/5817365/13683598/62fa9708-e6bc-11e5-97fb-51955235dad7.png)

dmehra commented:
Also, now when I load a juttle and click the Edit button, it continues showing `run mode: Path` and the original file path until I actually make an edit to the code, at which point it changes to `Local` and still `run mode: Path`, this doesn't seem quite right. I'd think we should switch display as soon as the Edit button is clicked, and also change mode to be not "Path" but something like "Edit" for clarity.

The editing problem described above severely limits my ability to try this out any further, will wait for a fix.

dmehra commented:
+1 on the working inputs, and the removed divider, much nicer.

The location of the Edit button is still not pretty, how about moving it to the top right corner of the code box, that way it will be visually under the expand handle, so buttons are sort of grouped together? Not a strong opinion.

davidvgalbraith commented:
Fixed the jumpy stuff.

davidvgalbraith commented:
Gave up on the height thing. Think we're gonna get rid of the part where it says the run mode so let's just leave that. Fixed everything else.

dmehra commented:
Request: let's not rebase until code review is +1 and you're ready to merge? Otherwise other people who are using the branch, let's say for testing, end up fighting with git after every pull. And for review it's easier if your changes that fix a specific issue are in a separate commit. At the end you'd rebase before merging, of course. How does that sound?

dmehra commented:
The editor is looking good! Should this one issue be filed separately: if the juttle in the editor has a syntax error, you can still click the Play button, at which point it's forever spinning and cannot be stopped (the only recovery is to refresh the page).

davidvgalbraith commented:
Yes, that one's separate.

dmehra commented:
Ok filed https://github.com/juttle/juttle-viewer/issues/54, I'm +1 on the editor.

mnibecker commented:
Actually it should be part of this pr. When we have an error, no bundle is set so you can't press play. Now that you can have a bundle and an error we'll need to make it so you can't press play when there's an error

mnibecker commented:
If your path is `/` and click view code it says `local` but you can't actually edit the code. To resolve this either:
1. Disable the view code button if mode isn't set
2. Redirect to `/?local` when path is `/`

I'm kind of leaning towards 1

mnibecker commented:
@davidvgalbraith added 9da1196 which makes the editor expand to the number of lines in the program

mnibecker commented on src/apps/actions.js:
```diff
@@ -18,10 +19,23 @@ export function newError(error) {
+    })
+    .catch(err => {
+        dispatch(fetchBundleError(bundleId, err, bundle));
+    });
```
instead of doing this lets add redux's [thunk middleware](https://github.com/gaearon/redux-thunk) and write the function like this:
```javascript
function promulgateBundle(bundle, bundleId) {
    return (dispatch, getState) => {
        const { juttleServiceHost } = getState();
        // rest of your stuff
    }
}
```

mnibecker commented on src/apps/reducers.js:
```diff
@@ -45,8 +45,8 @@ function bundleInfo(state = defaultBundleInfo, action) {
-                bundle: null,
-                inputs: null,
+                bundle: action.bundle,
+                inputs: state.inputs || [],
```
I think its fine we're doing this, we just need to resolve the issues this will call with the run button (ie the run button used to get disabled when there was error because the bundle was set to null).

mnibecker replied:
also perhaps we should just change inputs default to `[]` an ignore inputs on this one

mnibecker commented on src/apps/observers.js:
```diff
@@ -11,11 +11,7 @@ function runMode(store) {
-                store.dispatch(newBundle(bundleId, bundle, inputs));
-            })
-            .catch(err => { store.dispatch(fetchBundleError(bundleId, err)) })
+            return promulgateBundle(bundle, bundleId, juttleServiceHost, store.dispatch.bind(store));
```
I like this

mnibecker commented:
the runMode should get set to local when we switch `local` mode. I think this is causing some peculiar behavior in the app.

mnibecker commented on src/apps/components/juttle-viewer.js:
```diff
@@ -1,9 +1,54 @@
+        let isLocal = query.match(/\?local/);
+        this.state = {
+            readOnly: !isNewBundle && !isLocal
+        };
```
instead of doing this, lets set the runMode to local (via the reducer) and do this:
```javascript
this.state = { readOnly: !this.props.runMode.local }
```

mnibecker commented on src/apps/components/juttle-viewer.js:
```diff
@@ -1,9 +1,54 @@
+        </button> : null;
+        return (
+            <div className="juttle-editor">
+                {editButton}
```
lines 15-42 are what we use [react-router](https://github.com/reactjs/react-router) for.

It can be written as such:
```jsx
handleClick() {
    if (this.props.bundle) {
        localStorage.setItem('program', this.props.bundle.program);
    }
}
 
render() {
    return (
        <Link to={{ pathname: '/', query: { local: true }}} onClick={this.handleClick} className="btn btn-default" />
    );
}
```

mnibecker commented on src/apps/components/run-header.js:
```diff
@@ -12,13 +14,21 @@ const ENTER_KEY = 13;
+            let bundle = {program: localProgram};
+            promulgateBundle(bundle, LOCAL_BUNDLE_ID, juttleServiceHost, dispatch);
+        }
+    }
```
this logic should be moved to `src/apps/observers.js`

(from line 27)
```javascript
else if (runMode.local) {
    bundleRecieved(LOCAL_BUNDLE_ID, localStorage.getItem('program');
}
```

mnibecker commented on test/apps/components/juttle-editor.spec.js:
```diff
@@ -0,0 +1,15 @@
+            <JuttleEditor />
+        ));
+
+        ReactDOM.unmountComponentAtNode(ReactDOM.findDOMNode(el).parentNode);
```
can we actually do a test here?

mnibecker commented on src/apps/actions.js:
```diff
@@ -18,10 +19,23 @@ export function newError(error) {
+
+    return api.describe(juttleServiceHost, bundle)
+    .then(inputs => {
+        dispatch(newBundle(bundleId, bundle, inputs));
```
Why do we need to set newBundle twice? Also setting the inputs to `[]` causes the old input values to get cleared, I don't think we should be doing this.

mnibecker commented on src/apps/components/juttle-editor.js:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    render() {
+        let program = this.props.bundle ? this.props.bundle.program : '';
```
the fact that we set the program to the bundle, which could potentially be a second old seems pretty rickety. Plus the `<AceEditor />` already keeps track of value state so now your syncing state in two different places and hoping things work out (This is causing the cursor skipping around error that I still experience from time to time). 

What I think we should do is set the value of the editor on `componentDidMount`: `editor.setValue(this.props.bundle)`.

Then, slap a `key` on the `JuttleViewer` class instance ie `<JuttleViewer key={bundleId} />`. This will ensure that whenever the bundle changes, `componentDidMount` will get called and we're properly setting the value of editor to the current bundle/program.

mnibecker commented on src/apps/components/juttle-editor.js:
```diff
@@ -0,0 +1,67 @@
+              theme="monokai"
+              width="100%"
+              mode="juttle"
+              onChange={this.onChange.bind(this)}
```
nit: instead of binding you use es6 fat arrow methods:
```javascript
onChange = (newValue) => {
    // blah blah
};
```

mnibecker commented on src/apps/components/juttle-viewer.js:
```diff
@@ -1,9 +1,43 @@
+        if (isNewBundle) {
+            props.dispatch(newBundle(LOCAL_BUNDLE_ID, {program: ''}, {}));
+        }
+    }
```
I'm not a fan of this. 
- If there is no bundle, the JuttleViewer should never render and the code button should be disable. 
- I worry this would put the application in a weird state where we're in path mode but the bundleId is set to `local_bundle_id` which should really only be the case if the runMode is local
- In general views should only be changing redux state as a result of actions from the user. This would make it difficult for a developer to find out where the action is getting called.

mnibecker replied:
Rethinking my last comments- if there is no runMode and a user clicks the code button, reroute to local and let the observer instantiate a new bundle

mnibecker commented on src/apps/components/juttle-editor.js:
```diff
@@ -0,0 +1,74 @@
+    onChange = (newValue) => {
+        if (!this.props.readOnly) {
+            this._updateProgram(newValue);
+            localStorage.setItem('program', newValue);
```
suggestion: In the spirit of trying to keep as much logic out of the views, perhaps it might be better for the redux update action to set localStorage

mnibecker commented on src/apps/components/run.js:
```diff
@@ -42,8 +43,13 @@ export class RunApp extends React.Component {
 
+        if (nextProps.error) {
+            this._viewStatusChange(ViewStatus.STOPPED);
+        }
```
whoa, if you want to make sure the `viewStatus` goes to `STOPPED` you should call `view.stop` and let view fire a `viewStatus` change.

mnibecker commented:
From my pov we're good on functionality and code.

We need some real tests. A couple of suggestions for tests:
- add a test to `observers.spec` to verify that runMode local fires `promulgateBundle` with the correct bundle and bundleId when the mode changes to `local`
- in `components/juttle-viewer.js` verify that props readOnly displays the edit button.

I don't have any ideas for testing `components/juttle-editor.js`.

rlgomes commented:
The editor would end up being a simple app test in juttle-engine that would verify you can edit an existing juttle and rerun it. I filed an issue for me to follow up as part of the next release: https://github.com/juttle/juttle-engine/issues/92

mnibecker commented:
+1
